### ISSUE TITLE
Fail to wrap region with '(' or '{' in c++ mode.

### ISSUE THREAD

**Issue by moyotar** [Fuco1/smartparens#840]
emacs version: 24.4.1
smartparens version: up-to-date

`sp-local-paris's` value in c++ mode.
![image](https://user-images.githubusercontent.com/22092607/34712272-36d85f78-f55d-11e7-8312-5b8bb90129df.png)

**Comment by moyotar** [comment]
@Fuco1 What's the reason?

**Comment by Fuco1** [comment]
I don't know, you did not really provide any information.  It would be cool if you can follow the template when filling the issue.

**Comment by moyotar** [comment]
# Expected behavior
Type `(` or `{` to wrap region in c/c++ mode.

# Actual behavior
In c/c++ mode, when I activating a region and typing '(' or '{',  the region will be deleted instead of wrapping.

# Environment & version information
SMP Debian 3.16.43
GNU Emacs 24.4.1
smartparens: up-to-date

**Comment by moyotar** [comment]
@Fuco1 Is it clear?

**Comment by Fuco1** [comment]
Yes, but it works for me, I can not reproduce this.

**Comment by moyotar** [comment]
OK. Could you tell me how to debug this?

**Comment by Fuco1** [comment]
Is it an option for you to update to emacs 25?  I think some handling of delete-selection mode might have changed.

**Comment by moyotar** [comment]
Yes. It works. But in `GNU Emacs 25.2.1`, It will auto active the region after I type a pair such as `(`, `{`. How to config to disable this action?

**Comment by Fuco1** [comment]
I don't understand what you mean by /auto active/.  Do you mean that the selection is not cancelled? (i.e. typing anything will then delete the region)?

**Comment by moyotar** [comment]
I mean that a region between the pair which I type will be active automatically.

**Comment by moyotar** [comment]
Still fail to wrap region with '(' or '{' in c++ mode. What's more, It works normally in both of `lua-mode `and `elisp-mode` .
@Fuco1 How to debug this?
# Environment & version information
SMP Debian 3.16.43
GNU Emacs 26.1
smartparens: up-to-date

**Comment by moyotar** [comment]
the value of `sp-local-pairs` in `c++-mode`. And the type '[' can wrap region. It's really strange.
![image](https://user-images.githubusercontent.com/22092607/40657261-c80fdab8-6379-11e8-90d5-e98f3d0d37e3.png)

**Comment by Fuco1** [comment]
Are you using spacemacs or something like that?  Can you try to start emacs with `-Q` switch (that disables loading your configuration), then load dash and smartparens manually and then try this again?  I can't reproduce this locally.

**Comment by moyotar** [comment]
No, I use original GNU emacs and config by myself. Following your advice, I start emacs with `-Q`, then load `dash` and 'smartparens' only.
![image](https://user-images.githubusercontent.com/22092607/40695163-837490b4-63f2-11e8-96b5-398b3967eec5.png)
Magically, It works normally in c or c++ mode.
It's unintelligible because all of the config about `smartparens`  in my `init-file` is the same as above.

**Comment by Fuco1** [comment]
So it is probably some `c++-mode` configuration.  You can try to add that to the `-Q` setup and see what happens, I suspect there might be some of the "electric" keys which are making this not work: https://www.gnu.org/software/emacs/manual/html_node/ccmode/Electric-Keys.html

**Comment by moyotar** [comment]
Unexpectedly, It still worked after adding the config of `c++-mode`.

**Comment by moyotar** [comment]
Yes! Just as you thought, It's because `c-electric-paren` and ` c-electric-brace`, both of which are defined in `c-mode-base-map`. It became normally after I closed this manually.

`(dolist (key '("(" ")" "{" "}"))                                                                                                                 
  (define-key c-mode-base-map (kbd key) nil)                                                                                                     
  )              `

Thanks a lot.

**Comment by Fuco1** [comment]
We can also try to add those into `sp--special-self-insert-commands` and see if that helps.  The variable hods some of the "electric" commands from other modes like tex or racket.

**Comment by kehw** [comment]
I have the same problem. @moyotar 's solution works for me. Thanks.

**Comment by moyotar** [comment]
@Fuco1 @kehw Because I need `c-electric-brace` and `c-electric-paren` to reindent line appropriately in cc-mode, I recovered `c-mode-base-map`. And I find it can still wrap region normally.  I think the recent commits of `smartparens` have fixed this.

**Comment by moyotar** [comment]
@kehw Your can have a test about my guess. Just recover `c-mode-base-map`, update your `smartparens` package,  reload your config and see if it can still work.

**Comment by kehw** [comment]
Hi @moyotar , I just upgraded `smartparens` to Version: 20180912.1750. It still does not work if I recover `c-mode-base-map`. I am using Emacs 26.1.

**Comment by Fuco1** [comment]
@kehw @moyotar what happens if you try to add this configuration?

```elisp
(push 'c-electric-paren sp--special-self-insert-commands)
(push 'c-electric-brace sp--special-self-insert-commands)
```

**Comment by jgarvin** [comment]
In trying to debug my other issue I uninstalled smartparens and reinstalled, then ran into this. @moyotar's solution works for me, @Fuco1's did not.
